ui/pages/permissions-connect/connect-page/utils.ts

```diff
@@ -69,3 +72,61 @@ export function getCaip25PermissionsResponse(
     },
   };
 }
+
+/**
+ * Gets the default accounts for the requested namespaces.
+ * We need at least one default per requested namespace
+ * if there are more explicitly requested accounts, use those instead
+ * for that namespace
+ *
+ * @param requestedNamespaces - The namespaces requested.
+ * @param supportedRequestedAccounts - The supported requested accounts.
+ * @param allAccounts - All available accounts.
+ */
+export function getDefaultAccounts(
+  requestedNamespaces: CaipNamespace[],
+  supportedRequestedAccounts: MergedInternalAccountWithCaipAccountId[],
+  allAccounts: MergedInternalAccountWithCaipAccountId[],
+): MergedInternalAccountWithCaipAccountId[] {
+  const defaultAccounts: MergedInternalAccountWithCaipAccountId[] = [];
+
+  supportedRequestedAccounts.forEach((account) => {
+    const {
+      chain: { namespace },
+    } = parseCaipAccountId(account.caipAccountId);
+    if (requestedNamespaces.includes(namespace)) {
+      defaultAccounts.push(account);
+    }
+  });
+
+  // sort accounts by lastSelected descending
+  const allAccountsSortedByLastSelected =
+    sortSelectedInternalAccounts(allAccounts);
+
+  requestedNamespaces.forEach((namespace) => {
+    const defaultAccountExistsForNamespace = defaultAccounts.find((account) => {
+      const {
+        chain: { namespace: accountNamespace },
+      } = parseCaipAccountId(account.caipAccountId);
+      return accountNamespace === namespace;
+    });
+
+    if (defaultAccountExistsForNamespace) {
+      return;
```

Comment:
Nit: if you use `for...of` rather than `.forEach` above, then this can change to `continue`. That might communicate better that we are skipping this namespace rather than returning `undefined` from the function (which is impossible, but still).

Reply:
https://github.com/MetaMask/metamask-extension/pull/29887/commits/07ba772787be86543b4368a178ed4d738948a9d9